### ISSUE TITLE
DML EP is not executing nodes, despite it registering as a provider successfully

### ISSUE THREAD

**Issue by david-macleod** [microsoft/onnxruntime#9719]
### Discussed in https://github.com/microsoft/onnxruntime/discussions/9677

<div type='discussions-op-text'>

<sup>Originally posted by **david-macleod** November  5, 2021</sup>
I am currently experiencing an issue where models are not run using the DML EP, despite using the `onnxruntime-directml` wheel and the session object returning DML as a registered provider.

```py
import onnxruntime as ort
import torch

model = torch.nn.Sequential(*[torch.nn.Linear(4096, 4096) for _ in range(4)])
x = torch.randn(1, 4096)

torch.onnx.export(torch.jit.script(model), args=x, f="tmp.onnx", input_names=["x"], opset_version=13, example_outputs=x)

options = ort.SessionOptions()
options.log_severity_level = 0
session = ort.InferenceSession("tmp.onnx", options, providers=["DmlExecutionProvider"])
print("EPs", session.get_providers())

for _ in range(10):
    session.run(None, input_feed={"x": x.numpy()})
```
Which includes the following in the logs
```
2021-11-05 08:20:54.3613101 [I:onnxruntime:, inference_session.cc:1242 onnxruntime::InferenceSession::Initialize] Initializing session.
2021-11-05 08:20:54.3614344 [I:onnxruntime:, inference_session.cc:1279 onnxruntime::InferenceSession::Initialize] Adding default CPU execution provider.
2021-11-05 08:20:54.3663319 [I:onnxruntime:, reshape_fusion.cc:42 onnxruntime::ReshapeFusion::ApplyImpl] Total fused reshape node count: 0
2021-11-05 08:20:54.3668477 [I:onnxruntime:, reshape_fusion.cc:42 onnxruntime::ReshapeFusion::ApplyImpl] Total fused reshape node count: 0
2021-11-05 08:20:54.3681419 [V:onnxruntime:, inference_session.cc:152 onnxruntime::`anonymous-namespace'::VerifyEachNodeIsAssignedToAnEp] Node placements
2021-11-05 08:20:54.3683129 [V:onnxruntime:, inference_session.cc:154 onnxruntime::`anonymous-namespace'::VerifyEachNodeIsAssignedToAnEp] All nodes have been placed on [CPUExecutionProvider].
```
but also
```
EPs ['DmlExecutionProvider', 'CPUExecutionProvider']
```
Can anyone suggest a fix for this? I am using `onnxruntime-directml==1.9.0` and `torch==1.10.0+cu113`
</div>

**Comment by fdwr** [comment]
@david-macleod : Alas, "The DirectML execution provider currently supports ONNX opset 12" ([doc](https://onnxruntime.ai/docs/execution-providers/DirectML-ExecutionProvider.html#support-coverage)), and your model is opset 13. We have pending work to update the DML EP to 15, but for now, does opset 12 work for you?

**Comment by david-macleod** [comment]
Thanks! I have tried opset 12 and it works for my most of my current use cases, is there any timescale for when support for 15 might be available?

**Comment by fdwr** [comment]
@david-macleod Glad to hear it works for most. Opset 15 wouldn't be until after Christmas. There's a few weeks of work, and many of us are disappearing during chunks of December for winter vacation. ❄📆

**Comment by david-macleod** [comment]
Ok great, thanks for the heads-up!

**Comment by david-macleod** [comment]
Has there been any movement on this?

I notice that some ops appear to have opset 13 listed as supported, but this doesn't seem to be a recent addition
https://github.com/microsoft/onnxruntime/blob/6db6ee5710be7d247e7eace007dbfca1917a4cf8/onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp#L634

**Comment by fdwr** [comment]
@david-macleod : Yeah, I added a number of operators to opset 13 (see below from Nov and Feb), but not the complete opset which needs DML API updates. Hence our documentation still states opset 12, because claiming support for opset 13 would be misleading as those ops would still fall back to the CPU (which can yield bad perf due to GPU<->CPU stalls). The remaining opset support will be out by ORT 1.12.

- Abs-13
- Add-13
- ArgMax-13
- ArgMin-13
- Cast-13
- Ceil-13
- Clip-13
- Concat-13
- Constant-13
- DepthToSpace-13
- DequantizeLinear-13
- Div-13
- Equal-13
- Erf-13
- Exp-13
- Expand-13
- Flatten-13
- Floor-13
- Gather-13
- GatherElements-13
- GatherND-13
- Gemm-13
- Greater-13
- Identity-13
- IsNaN-13
- LRN-13
- Less-13
- Log-13
- MatMul-13
- Max-13
- Mean-13
- MeanVarianceNormalization-13
- Min-13
- Mod-13
- Mul-13
- Neg-13
- Pad-13
- Pow-13
- QuantizeLinear-13
- Reciprocal-13
- ReduceL1-13
- ReduceL2-13
- ReduceLogSum-13
- ReduceLogSumExp-13
- ReduceMax-13
- ReduceMean-13
- ReduceMin-13
- ReduceProd-13
- ReduceSum-13
- ReduceSumSquare-13
- Relu-13
- Reshape-13
- Reshape-13
- Resize-13
- Scatter-13
- ScatterElements-13
- ScatterND-13
- Sigmoid-13
- Sign-13
- Slice-13
- SpaceToDepth-13
- Split-13
- Sqrt-13
- Squeeze-13
- Sub-13
- Sum-13
- Tanh-13
- Tile-13
- Transpose-13
- Unsqueeze-13
- Upsample-13

**Comment by carsonswope** [comment]
Hi @fdwr - would this explain the errors that I'm seeing trying to run an ONNX model w/ opset 15 and a Div node (w/ DML EP)?

```
Error running model inference: Non-zero status code returned while running Div node. Name:'Div_3' Status Message: D:\a\_work\1\s\include\onnxruntime\core/common/logging/logging.h:289 onnxruntime::logging::LoggingManager::DefaultLogger Attempt to use DefaultLogger but none has been registered.
```

**Comment by fdwr** [comment]
@carsonswope What ORT version? 1.12 in current master?

Your model should get CPU fallback for [Div opset 14](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Div) (there is no Div-15, meaning it resolves to Div-14) in absence of a higher priority EP registration. The error message is perplexing because the DML EP barely uses the logger, and certainly not the DML EP's elementwise division kernel. Do you get the same error running just CPU?

**Comment by fdwr** [comment]
@david-macleod The ORT 1.12 release is imminent, supporting 13-15.

**Comment by carsonswope** [comment]
@fdwr it was on ORT version 1.11.1. Just tested, works as expected with the 1.12 release candidate.

I guess the CPU fallback wasn't working correctly. Happy to provide more info if you want to debug it, but if it's fixed going forward I will just use 1.12 when it's released.

Thanks for the example code of how to use DirectML w/ ORT by the way (https://github.com/fdwr/OnnxRuntimeDirectMLEPSample). It's been super helpful for me :)

**Comment by fdwr** [comment]
@carsonswope 
> works as expected with the 1.12 release candidate

Glad to hear.

>Thanks for the example code of how to use DirectML w/ ORT by the way

Welcome. Note it's not an *official* blessed sample (I believe @martinb35 has work lined up for the official repo), but I just needed something *myself* to debug it too 😋.

**Comment by fdwr** [comment]
@david-macleod I'm closing this, as opsets 13-15 are now supported in the DML EP (in master and the release candidate, with the official release very soon).
